pkg/merkletree/tree_test.go

```diff
@@ -0,0 +1,357 @@
+			[]byte("002"),
+			[]byte("003"),
+			[]byte("004"),
+			[]byte("005"),
+			[]byte("006"),
+			[]byte("007"),
+			[]byte("008"),
+		}
+
+		tree, err := NewTree(leaves, mockHashFunc)
+		require.NoError(t, err)
+		require.Equal(t, [][]byte{[]byte("002"), []byte("007"), []byte("026")}, tree.GetProof([]byte("001")))
+		require.Equal(t, [][]byte{[]byte("001"), []byte("007"), []byte("026")}, tree.GetProof([]byte("002")))
+		require.Equal(t, [][]byte{[]byte("004"), []byte("003"), []byte("026")}, tree.GetProof([]byte("003")))
+		require.Equal(t, [][]byte{[]byte("003"), []byte("003"), []byte("026")}, tree.GetProof([]byte("004")))
+
+		require.Equal(t, [][]byte{[]byte("006"), []byte("015"), []byte("010")}, tree.GetProof([]byte("005")))
+		require.Equal(t, [][]byte{[]byte("005"), []byte("015"), []byte("010")}, tree.GetProof([]byte("006")))
+		require.Equal(t, [][]byte{[]byte("008"), []byte("011"), []byte("010")}, tree.GetProof([]byte("007")))
+		require.Equal(t, [][]byte{[]byte("007"), []byte("011"), []byte("010")}, tree.GetProof([]byte("008")))
+	})
+
+	t.Run("not found", func(t *testing.T) {
+		t.Parallel()
+		leaves := [][]byte{
+			[]byte("001"),
+			[]byte("002"),
+			[]byte("003"),
+			[]byte("004"),
+			[]byte("005"),
+		}
+
+		tree, err := NewTree(leaves, mockHashFunc)
+		require.NoError(t, err)
+
+		require.Nil(t, tree.GetProof([]byte("006")))
+		require.Len(t, tree.GetProof([]byte("006")), 0)
+	})
+}
+
+func TestVerifyProof(t *testing.T) {
+	t.Parallel()
+	t.Run("correct proof", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("036")
+		proof := [][]byte{[]byte("002"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.True(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+
+	t.Run("wrong root", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("035")
+		proof := [][]byte{[]byte("002"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.False(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+
+	t.Run("wrong proof", func(t *testing.T) {
+		t.Parallel()
+		root := []byte("036")
+		proof := [][]byte{[]byte("001"), []byte("007"), []byte("026")}
+		leaf := []byte("001")
+		require.False(t, VerifyProof(root, proof, leaf, mockHashFunc))
+	})
+}
+
+func TestProperties(t *testing.T) {
+	t.Parallel()
+
+	// We test the properties in a bunch of hash functions to make sure the
+	// kind of hash function has no influence on properties.
+	hashFuncs := []func(...[]byte) []byte{
+		nil,
+		mockHashFunc,
+		crypto.Keccak256,
+		crypto.Keccak512,
+		func(b ...[]byte) []byte {
+			h := sha3.Sum224(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum256(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum384(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha3.Sum512(b[0])
+			return h[:]
+		},
+		func(b ...[]byte) []byte {
+			h := sha1.Sum(b[0])
+			return h[:]
+		},
+	}
+
+	t.Run("tree holds merkle tree property", func(t *testing.T) {
```

Comment:
this tests if the tree holds the merkle tree property: the hashing of the leaves leads to the merkle root